### PR TITLE
feat: Check if VM supports TSC

### PR DESCRIPTION
**Pull request by ariesdevil**
Invariant TSC could make sure TSC got synced among multiple CPUs, so we check CPUID first, `0x80000007` is a specific CPUID leaf used to query the advanced power management features of a processor, including support for an invariant TSC.

In some VMs, the Max Extended Function in CPUID is < 0x80000007, we should enable TSC if the system clock source is TSC.

FYI: [CPUID explorer](http://www.flounder.com/cpuid_explorer2.htm#CPUID(0x80000007))

### CODE REVIEW

File: src/tsc_now.rs
Comment (tisonkun) [inline src/tsc_now.rs:115]: Perhaps tolerate all errors (just `continue`)?

**Review by tisonkun — COMMENTED** [review]
Thanks for your contribution. Generally looks good.

Waiting for other maintainer's review.

File: src/tsc_now.rs
Comment (ariesdevil) [reply src/tsc_now.rs:115]: other errors should return false i think.

Comment (tisonkun) [reply src/tsc_now.rs:115]: Why? You test either file exists and if any one of them exist and contains "TSC", it means "TSC" is valid.

In other words, what kind of other errors are possible and how they indicate an early false?

Comment (ariesdevil) [reply src/tsc_now.rs:115]: The current implementation logic is the same as before(`clock_source.map(|s| s.contains("tsc")).unwrap_or(false)`). Other errors, such as `PermissionDenied` should return false.

Comment (tisonkun) [reply src/tsc_now.rs:115]: But perviously we check only one file, and now we check two. The previous logic is the same as:

```rust
        let clock_sources = [
            "/sys/devices/system/clocksource/clocksource0/available_clocksource",
        ];

        for path in clock_sources {
            match read_to_string(path) {
                Ok(content) => {
                    if content.contains("tsc") {
                        return true;
                    }
                }
                Err(e) => continue,
            }
        }

        false
```

Comment (ariesdevil) [reply src/tsc_now.rs:115]: The previous implementation was not entirely correct, we should check `current_clocksource` first, if `current_clocksource` is not provided, we fallback check `available`.

`current_clocksource` is the clocksource system specified by the sysadmin, if they won't let user use `tsc`, they may set `current_clocksource` to such as `hpet`.

Comment (tisonkun) [reply src/tsc_now.rs:115]: This is my expected code ...

```rust
        let clock_sources = [
            "/sys/devices/system/clocksource/clocksource0/current_clocksource",
            "/sys/devices/system/clocksource/clocksource0/available_clocksource",
        ];

        for path in clock_sources {
            match read_to_string(path) {
                Ok(content) => {
                    if content.contains("tsc") {
                        return true;
                    }
                }
                Err(e) => continue,
            }
        }

        false
```

The difference is that, no matter `current_clocksource` is missing, cannot access, and other errors, we can try `available_clocksource`.

I don't find the reason to refuse test `available_clocksource` when read `current_clocksource` returns some errors.

Comment (ariesdevil) [reply src/tsc_now.rs:115]: I don't think so.

Only read `available_clocksource ` if `current_clocksource` doesn't exist, and follow `current_clocksource` as long as `current_clocksource` exists, regardless of what `available_clocksource` provides.

Imagine a scenario where the administrator has set `hpet` in `current_clocksource`, but `tsc` is provided in `available_clocksource `. Obviously, the administrator does not want to use `tsc`, and if we reads `available_clocksource `, it will be difficult to troubleshoot the problem if needed because it does not meet the administrator's expectations.

Comment (tisonkun) [reply src/tsc_now.rs:115]: You current code doesn't work in the way you describe. It will read `available_clocksource` if `current_clocksource` exists but doesn't contain `tsc`.

To implement the manner your describe, the code is like:

```rust
{
    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
        return content.contains("tsc")
    }

    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
        return content.contains("tsc")
    }

    false
}
```

Comment (ariesdevil) [reply src/tsc_now.rs:115]: `available_clocksource ` is just a template file, we fallback to this iff `current_clocksource` not exists(rarely)

Comment (ariesdevil) [reply src/tsc_now.rs:115]: > You current code doesn't work in the way you describe. It will read `available_clocksource` if `current_clocksource` exists but doesn't contain `tsc`.
> 
> To implement the manner your describe, the code is like:
> 
> ```rust
> {
>     if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
>         return content.contains("tsc")
>     }
> 
>     if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
>         return content.contains("tsc")
>     }
> 
>     false
> }
> ```

Correct! let me fix it.

Comment (tisonkun) [reply src/tsc_now.rs:115]: Or if you consider that a `current_clocksource` that permission deny should indicate a conservative resolution to return false (reject to use TSC), write:

```rust
{
    match read_to_string("/sys/devices/system/clocksource/clocksource0/current_clocksource") {
        Ok(content) => return content.contains("tsc")
        Err(e) if e.kind() == ErrorKind::NotFound => {}
        Err(_) => return false,
    }

    if let Ok(content) = read_to_string("/sys/devices/system/clocksource/clocksource0/available_clocksource") {
        return content.contains("tsc")
    }

    false
}
```

Comment (ariesdevil) [reply src/tsc_now.rs]: Fixed, thx~

**Review by tisonkun — APPROVED** [review]
@andylokandy @zhongzc I'm yet to have time check if `has_invariant_tsc` work reasonably. Would you take a look?

**Comment by tisonkun** [comment]
@ariesdevil I noticed that we don't release this patch. Do you need a new version for use?

**Pull request merged**
